calico-vpp-agent/routing/bgp_watcher.go

```diff
@@ -267,6 +269,180 @@ func (w *Server) startBGPMonitoring() (func(), error) {
 	return stopFunc, err
 }
 
+func (w *Server) NewBGPPolicyAndAssignment(name string, rules []calicov3.BGPFilterRuleV4, neighborName string, dir bgpapi.PolicyDirection) (*watchers.BGPPolicyAndAssignment, error) {
+	pol := &bgpapi.Policy{Name: name}
+	for _, rule := range rules {
+		routeAction := bgpapi.RouteAction_ACCEPT
+		if rule.Action == calicov3.Reject {
+			routeAction = bgpapi.RouteAction_REJECT
+		} else if rule.Action != calicov3.Accept {
+			w.log.Errorf("action %s not supported", rule.Action)
+		}
+
+		var matchSetType bgpapi.MatchSet_Type
+		var minMask, maxMask uint32
+		if rule.MatchOperator == calicov3.In || rule.MatchOperator == calicov3.NotIn {
+			_, subnet, err := net.ParseCIDR(rule.CIDR)
+			if err != nil {
+				w.log.Error(err)
+			}
+			ones, bits := subnet.Mask.Size()
+			minMask = uint32(ones)
+			maxMask = uint32(bits)
+			if rule.MatchOperator == calicov3.In {
+				matchSetType = bgpapi.MatchSet_ANY // any and all are same in our case as we have only one member of the defined set
+			} else {
+				matchSetType = bgpapi.MatchSet_INVERT
+			}
+		} else {
+			// mask is zero
+			if rule.MatchOperator == calicov3.Equal {
+				matchSetType = bgpapi.MatchSet_ANY
+			} else {
+				matchSetType = bgpapi.MatchSet_INVERT
+			}
+		}
+
+		prefixName := rule.CIDR + "prefix" + fmt.Sprint(minMask) + fmt.Sprint(maxMask) // this name should be unique
+		err := w.BGPServer.AddDefinedSet(context.Background(), &bgpapi.AddDefinedSetRequest{
+			DefinedSet: &bgpapi.DefinedSet{
+				DefinedType: bgpapi.DefinedType_PREFIX,
+				Name:        prefixName,
+				Prefixes:    []*bgpapi.Prefix{{IpPrefix: rule.CIDR, MaskLengthMin: minMask, MaskLengthMax: maxMask}}},
+		})
```

Comment:
Maybe we can move this part in the caller function `createFilterPolicy()` so that every `w.BGPServer()` happen at the right place. 
This would have the benefit of not creating leftover state when part of the parsing fails (i.e. some rules have unsupported types)

Reply:
Also are we sure we do not need to delete this definedSet before deleting the Peer in goBGP ?
If yes, we should probably add a comment that explains it